components/brave_ads/core/browser/service/oblivious_http_key_config.cc

```diff
@@ -0,0 +1,201 @@
+/* Copyright (c) 2025 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_ads/core/browser/service/oblivious_http_key_config.h"
+
+#include <algorithm>
+#include <utility>
+
+#include "base/base64.h"
+#include "base/check.h"
+#include "base/functional/bind.h"
+#include "base/location.h"
+#include "base/logging.h"
+#include "base/rand_util.h"
+#include "base/time/time.h"
+#include "brave/components/brave_ads/core/browser/service/network_client_util.h"
+#include "brave/components/brave_ads/core/browser/service/oblivious_http_feature.h"
+#include "brave/components/brave_ads/core/public/prefs/pref_names.h"
+#include "components/prefs/pref_service.h"
+#include "net/http/http_request_headers.h"
+#include "net/http/http_status_code.h"
+#include "services/network/public/cpp/resource_request.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "services/network/public/mojom/url_response_head.mojom.h"
+
+namespace brave_ads {
+
+namespace {
+constexpr int kMaxFetchJitterInSeconds = 60;
+}  // namespace
+
+ObliviousHttpKeyConfig::ObliviousHttpKeyConfig(
+    PrefService* local_state,
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory,
+    GURL key_config_url)
+    : local_state_(local_state),
+      url_loader_factory_(std::move(url_loader_factory)),
+      key_config_url_(std::move(key_config_url)) {
+  CHECK(local_state_);
+}
+
+ObliviousHttpKeyConfig::~ObliviousHttpKeyConfig() = default;
+
+void ObliviousHttpKeyConfig::MaybeFetch() {
+  if (HasExpired()) {
+    VLOG(6) << "OHTTP key config has expired";
+    return Fetch();
+  }
+
+  FetchAfter(kOhttpKeyConfigExpiresAfter.Get());
+}
+
+void ObliviousHttpKeyConfig::InvalidateAndFetch() {
+  Invalidate();
+  Fetch();
+}
+
+std::optional<std::string> ObliviousHttpKeyConfig::Get() const {
+  if (!local_state_->HasPrefPath(prefs::kObliviousHttpKeyConfig)) {
+    return std::nullopt;
+  }
+
+  const auto key_config = base::Base64Decode(
+      local_state_->GetString(prefs::kObliviousHttpKeyConfig));
+  if (!key_config) {
+    return std::nullopt;
+  }
+
+  return std::string(key_config->cbegin(), key_config->cend());
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+base::TimeDelta ObliviousHttpKeyConfig::ExpiresAfter() const {
+  if (HasExpired()) {
+    // Already expired.
+    return base::TimeDelta::Min();
+  }
+
+  return local_state_->GetTime(prefs::kObliviousHttpKeyConfigExpiresAt) -
```

Comment:
This may or may not be a problem in practice, but there's a potential race condition here. We're checking `now >= expiresAt` first and then later we do `expiresAt - now`. We may therefore end up returning a negative number if it's close enough.

Reply:
`now` is a local variable whose value does not change after it is assigned.

Reply:
> now is a local variable whose value does not change after it is assigned.
You're right, I missed that change.

Reply:
:-), thanks much appreciated